File: neuron_db/turn.py
"""Conversation routing over a Neuron. Deterministic: a statement is stored and
acknowledged, a question is answered from memory or with an honest "i don't know".
Native arithmetic for +,-,*,/. (The trained gary-neuron adder model is a separate
project and is NOT used here -- neuron-db is model-free by design.)
"""
from __future__ import annotations
import re
from .neuron import Neuron, _content, _words, _isnum, COMMAND

QWORDS = {"what", "whats", "where", "who", "when", "how", "which"}
YNWORDS = {"am", "is", "are", "do", "does", "did", "can", "could", "will", "would", "was", "were", "have", "has"}
GREET = re.compile(r"^(hi+|hey+|hello+|yo|sup|howdy|good (morning|afternoon|evening))\b[\s!,.?]*$", re.I)
LAUGH = re.compile(r"^(lo+l+|lmao+|haha+|hehe+)\b", re.I)
THANKS = re.compile(r"^(thanks|thank you|thx|ty)\b", re.I)
BYE = re.compile(r"^(bye+|goodbye|later|gtg|good ?night)\b", re.I)
HOWRU = re.compile(r"how('s| is| are)? (it going|you( doing)?|things)", re.I)
SELFQ = re.compile(r"real\s+person|human|alive|robot|an?\s+ai\b|sentient|a\s+machine", re.I)
MATH = re.compile(r"(-?\d{1,12})\s*([+\-*/])\s*(-?\d{1,12})")
GREETS = ["hey. tell me things; i'll remember them.", "hi. what should i remember?",
          "ready. give me facts, ask them back later.", "memory online. go ahead."]
CHATR = ["noted.", "okay.", "got it.", "if it matters, say it like a fact -- i'll keep it."]
IDK = "i don't know right now."


def turn(n: Neuron, u: str) -> dict:
    """Returns {'reply', 'kind', 'wrote'}. kind in: ack recall idk smalltalk math self."""
    u = u.strip()
    uw = _words(u)
    first = u.split()[0].lower().strip("?,!") if u else ""
    questionish = ("?" in u) or (first in QWORDS)
    about = (uw & {"your", "you", "yours"}) and not (uw & {"my", "i", "im", "mine"})
    yn = (first in YNWORDS) or ("yes or no" in u.lower())
    if yn: questionish = True

    if about and re.search(r"what('s| is) your name|who are you", u.lower()):
        return _r("i'm a neuron -- an associative memory. you're the one i remember things about.", "self")
    if about and SELFQ.search(u):
        return _r("i'm a tiny memory, not a person. but i'll remember what you tell me.", "self")
    if GREET.match(u): return _r(GREETS[n.fact_count % len(GREETS)], "smalltalk")
    if LAUGH.match(u): return _r("glad that landed.", "smalltalk")
    if THANKS.match(u): return _r("anytime.", "smalltalk")
    if BYE.match(u): return _r("later. i'll remember.", "smalltalk")
    if questionish and HOWRU.search(u): return _r("running fine -- all of it yours.", "smalltalk")
    if COMMAND.match(u) and not (uw & {"my", "i", "im", "mine"}):
        return _r("i can't do that -- i remember facts and do arithmetic.", "smalltalk")

    # native arithmetic: + - * /  (word forms too)
    mm = u
    for word, sym in (("plus", "+"), ("minus", "-"), ("times", "*"), ("divided by", "/")):
        mm = re.sub(r"\b" + word + r"\b", sym, mm, flags=re.I)
    m = MATH.search(mm)
    if m and not re.search(r"=\s*-?\d", mm):
        a, op, b = int(m.group(1)), m.group(2), int(m.group(3))
        r = {"+": a + b, "-": a - b, "*": a * b, "/": (round(a / b, 6) if b else None)}.get(op)
        if r is not None:
            return _r(f"{a} {op} {b} = {r}", "math")

    if questionish:
        cw = _content(u)
        if not cw and re.search(r"\bwho am i\b", u.lower()): cw = {"name"}
        hit = n.recall(cw) if cw else None
        if yn and cw:
            if hit and hit["coverage"] >= 0.99: return _r("yes.", "recall")
            if hit: return _r(f'hmm -- what i remember is: {hit["fact"]}', "recall")
            return _r("not that i know.", "recall")
        if hit:
            v = hit["value"]
            conf = _isnum(v) or (v[:1].isupper() and len(hit["fact"].split()) <= 6)
            if conf and not hit["echo"]: return _r(f"{v}.", "recall")
            if hit["echo"] or len(hit["fact"].split()) > 6:
                q = hit["fact"].strip()
                return _r(f'what i remember: "{q if len(q) <= 140 else q[:137] + chr(8230)}"', "recall")
            return _r(f"{v}.", "recall")
        if cw:
            return _r(IDK, "idk")

    wrote = n.observe(u)
    if wrote:
        val = wrote[-1]["v"]
        r = f"noted -- {len(wrote)} facts." if len(wrote) > 1 else [f"got it -- {val}.", f"noted -- {val}.", f"okay -- {val}."][n.fact_count % 3]
        return {"reply": r, "kind": "ack", "wrote": len(wrote)}
    return _r(CHATR[len(u) % len(CHATR)], "smalltalk")


def _r(reply: str, kind: str) -> dict:
    return {"reply": reply, "kind": kind, "wrote": 0}

File: rust/neuron-core/src/turn.rs
//! turn(): deterministic conversation routing over a Neuron. A statement is stored and
//! acknowledged; a question is answered from memory or with an honest "i don't know";
//! arithmetic (+ - * /, word forms too) is evaluated. std-only, no regex crate, wasm-safe.
//! Faithful port of neuron_db/turn.py.

use crate::{Neuron, content, words, is_num};

pub struct Turn { pub reply: String, pub kind: String, pub wrote: usize }
fn rr(reply: &str, kind: &str) -> Turn { Turn { reply: reply.into(), kind: kind.into(), wrote: 0 } }

const QWORDS: [&str; 7] = ["what","whats","where","who","when","how","which"];
const YNWORDS: [&str; 14] = ["am","is","are","do","does","did","can","could","will","would","was","were","have","has"];
const COMMANDS: [&str; 16] = ["count","explain","list","show","give","make","stop","repeat","calculate","compute","draw","sing","translate","define","describe","summarize"];
const GREET_ROOTS: [&str; 6] = ["hi","hey","hello","yo","sup","howdy"];
const GREETS: [&str; 4] = ["hey. tell me things; i'll remember them.","hi. what should i remember?",
    "ready. give me facts, ask them back later.","memory online. go ahead."];
const CHATR: [&str; 4] = ["noted.","okay.","got it.","if it matters, say it like a fact -- i'll keep it."];
const IDK: &str = "i don't know right now.";

fn lc(s: &str) -> String { s.to_lowercase() }
fn first_word(u: &str) -> String {
    u.split_whitespace().next().unwrap_or("").trim_matches(|c: char| "?,!".contains(c)).to_lowercase()
}
// squeeze a trailing run of the last char to one ("hiii"->"hi", "heyyy"->"hey")
fn squeeze_tail(s: &str) -> String {
    let mut v: Vec<char> = s.chars().collect();
    while v.len() > 2 && v[v.len()-1] == v[v.len()-2] { v.pop(); }
    v.into_iter().collect()
}
fn is_greet(u: &str) -> bool {
    let t = lc(u.trim());
    if t.starts_with("good morning") || t.starts_with("good afternoon") || t.starts_with("good evening") { return true; }
    let head = squeeze_tail(&first_word(&t));
    GREET_ROOTS.contains(&head.as_str())
}
fn starts_any(u: &str, pres: &[&str]) -> bool { let t = lc(u); pres.iter().any(|p| t.starts_with(p)) }
fn contains_any(u: &str, subs: &[&str]) -> bool { let t = lc(u); subs.iter().any(|s| t.contains(s)) }

// arithmetic: a space-delimited "<int> <op> <int>" (so dates like 2026-06-13 are NOT math)
fn find_math(s: &str) -> Option<(i64, char, i64)> {
    let mut t = format!(" {} ", lc(s));
    for (w, sym) in [("divided by"," / "),("plus"," + "),("minus"," - "),("times"," * ")] {
        t = t.replace(&format!(" {} ", w), sym);
    }
    let toks: Vec<&str> = t.split_whitespace().collect();
    for w in toks.windows(3) {
        if let (Ok(a), Ok(b)) = (w[0].parse::<i64>(), w[2].parse::<i64>()) {
            if w[1].len() == 1 {
                let op = w[1].chars().next().unwrap();
                if "+-*/".contains(op) { return Some((a, op, b)); }
            }
        }
    }
    None
}

pub fn turn(n: &mut Neuron, u: &str) -> Turn {
    let u = u.trim();
    let uw = words(u);
    let first = first_word(u);
    let mut questionish = u.contains('?') || QWORDS.contains(&first.as_str());
    let about = (uw.contains("your") || uw.contains("you") || uw.contains("yours"))
        && !(uw.contains("my") || uw.contains("i") || uw.contains("im") || uw.contains("mine"));
    let yn = YNWORDS.contains(&first.as_str()) || lc(u).contains("yes or no");
    if yn { questionish = true; }

    if about && contains_any(u, &["what's your name","what is your name","whats your name","who are you"]) {
        return rr("i'm a neuron -- an associative memory. you're the one i remember things about.", "self");
    }
    if about && contains_any(u, &["real person","human","alive","robot"," ai ","sentient","a machine"]) {
        return rr("i'm a tiny memory, not a person. but i'll remember what you tell me.", "self");
    }
    if is_greet(u) { return rr(GREETS[n.fact_count() % GREETS.len()], "smalltalk"); }
    if starts_any(u, &["lol","lmao","haha","hehe"]) { return rr("glad that landed.", "smalltalk"); }
    if starts_any(u, &["thanks","thank you","thx","ty "]) || lc(u) == "ty" { return rr("anytime.", "smalltalk"); }
    if starts_any(u, &["bye","goodbye","later","gtg","good night","goodnight"]) { return rr("later. i'll remember.", "smalltalk"); }
    if questionish && contains_any(u, &["how's it going","how is it going","how are you","how you doing","how are things"]) {
        return rr("running fine -- all of it yours.", "smalltalk");
    }
    if COMMANDS.contains(&first.as_str()) && !(uw.contains("my")||uw.contains("i")||uw.contains("im")||uw.contains("mine")) {
        return rr("i can't do that -- i remember facts and do arithmetic.", "smalltalk");
    }

    if !lc(u).contains('=') {
        if let Some((a, op, b)) = find_math(u) {
            let reply = match op {
                '+' => Some(format!("{} + {} = {}", a, b, a + b)),
                '-' => Some(format!("{} - {} = {}", a, b, a - b)),
                '*' => Some(format!("{} * {} = {}", a, b, a * b)),
                '/' => if b != 0 { Some(format!("{} / {} = {}", a, b, (((a as f64 / b as f64) * 1e6).round() / 1e6))) } else { None },
                _ => None,
            };
            if let Some(rep) = reply { return rr(&rep, "math"); }
        }
    }

    if questionish {
        let cw = content(u);
        let hit = if !cw.is_empty() { n.recall(u) }
                  else if lc(u).contains("who am i") { n.recall("name") }
                  else { None };
        if yn && !cw.is_empty() {
            return match &hit {
                Some(h) if h.coverage >= 0.99 => rr("yes.", "recall"),
                Some(h) => rr(&format!("hmm -- what i remember is: {}", h.fact), "recall"),
                None => rr("not that i know.", "recall"),
            };
        }
        if let Some(h) = &hit {
            let v = &h.value;
            let cap = v.chars().next().map_or(false, |c| c.is_uppercase());
            let conf = is_num(v) || (cap && h.fact.split_whitespace().count() <= 6);
            if conf && !h.echo { return rr(&format!("{}.", v), "recall"); }
            if h.echo || h.fact.split_whitespace().count() > 6 {
                let q = h.fact.trim();
                let shown = if q.chars().count() <= 140 { q.to_string() }
                            else { format!("{}\u{2026}", q.chars().take(137).collect::<String>()) };
                return rr(&format!("what i remember: \"{}\"", shown), "recall");
            }
            return rr(&format!("{}.", v), "recall");
        }
        if !cw.is_empty() { return rr(IDK, "idk"); }
    }

    let wrote = n.observe(u);
    if wrote > 0 {
        let val = n.episodes.last().map(|e| e.v.clone()).unwrap_or_default();
        let reply = if wrote > 1 { format!("noted -- {} facts.", wrote) }
            else { match n.fact_count() % 3 { 0 => format!("got it -- {}.", val), 1 => format!("noted -- {}.", val), _ => format!("okay -- {}.", val) } };
        return Turn { reply, kind: "ack".into(), wrote };
    }
    rr(CHATR[u.len() % CHATR.len()], "smalltalk")
}
